kevm-pyk/src/kevm_pyk/kproj/evm-semantics/lemmas/evm-int-simplification.k

```diff
@@ -57,31 +50,92 @@ module EVM-INT-SIMPLIFICATION-COMMON
   // #asWord
   // ###########################################################################
 
-  rule [asWord-lb]: 0 <=Int #asWord(_WS)     => true [simplification, smt-lemma]
-  rule [asWord-ub]: #asWord(_WS) <Int pow256 => true [simplification, smt-lemma]
-
-  rule [asWord-lt]:
-    #asWord ( BA ) <Int X => true
-    requires 2 ^Int ( 8 *Int lengthBytes(BA) ) <=Int X
-    [concrete(X), simplification]
-
-  rule [asWord-lt-concat]:
-    #asWord ( BA1 +Bytes BA2 ) <Int X => #asWord ( BA1 ) <Int ( X -Int #asWord ( BA2 ) ) /Int ( 2 ^Int ( 8 *Int lengthBytes ( BA2 ) ) )
-    requires ( X -Int #asWord ( BA2 ) ) modInt ( 2 ^Int ( 8 *Int lengthBytes ( BA2 ) ) ) ==Int 0
-    [concrete(BA2, X), simplification, preserves-definedness]
-
-  rule [asWord-lt-range]:
-    #asWord ( #range ( _:Bytes, S, W ) ) <Int X => true
-    requires 0 <=Int S andBool 0 <=Int W
-     andBool 2 ^Int ( 8 *Int W ) <=Int X
-    [simplification, concrete(S, W, X), preserves-definedness]
-
-  rule [lt-asWord-range]:
-    X <Int #asWord ( #range ( B:Bytes, S, W ) ) => X <Int #asWord ( B )
+  // Unification: two `#asWord`s
+  rule [asWord-eq-asWord]:
+    #asWord ( B1 ) ==Int #asWord ( B2 ) => #buf ( 32 -Int lengthBytes(B1), 0 ) +Bytes B1 ==K #buf ( 32 -Int lengthBytes(B2), 0 ) +Bytes B2
+    requires lengthBytes(B1) <=Int 32 andBool lengthBytes(B2) <=Int 32
+    [simplification, preserves-definedness]
+
+  // Unification: two `#asWord`s
+  rule [asWord-eq-asWord-ml]:
+    { #asWord ( B1 ) #Equals #asWord ( B2 ) } => { true #Equals #asWord ( B1 ) ==Int #asWord ( B2 ) }
+    [simplification, preserves-definedness]
+
+  // Unification: `#asWord` and concrete number
+  rule [asWord-eq-num]:
+    #asWord ( B:Bytes ) ==Int X:Int =>
+      0 <=Int X andBool X <Int 2 ^Int (8 *Int minInt(32, lengthBytes(B)))
+                andBool (
+                  ( lengthBytes(B) <=Int 32 andBool #buf ( 32 -Int lengthBytes(B), 0 ) +Bytes B ==K #buf ( 32, X ) )
+                    orBool
+                  ( lengthBytes(B) >Int 32 andBool #range ( B, lengthBytes(B) -Int 32, 32 ) ==K #buf ( 32, X ) )
+                )
+    [simplification, concrete(X), comm, preserves-definedness]
+
+  // Unification: `#asWord` and concrete number
+  rule [asWord-eq-num-ml-l]: { #asWord ( B:Bytes ) #Equals X:Int } => { true #Equals #asWord ( B:Bytes ) ==K X:Int } [simplification, concrete(X), preserves-definedness]
+  rule [asWord-eq-num-ml-r]: { X:Int #Equals #asWord ( B:Bytes ) } => { true #Equals #asWord ( B:Bytes ) ==K X:Int } [simplification, concrete(X), preserves-definedness]
```

Comment:
Required for now, will go away once booster takes over.

Reply:
Can probably use `comm` attribute to only have one of these rules.

Reply:
`comm` doesn't work on ML-level rules, sadly.